mergify_engine/eventlogs.py

```diff
@@ -0,0 +1,155 @@
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import datetime
+import itertools
+import typing
+
+import daiquiri
+import msgpack
+
+from mergify_engine import date
+from mergify_engine import github_types
+from mergify_engine import signals
+from mergify_engine.dashboard import subscription
+
+
+if typing.TYPE_CHECKING:
+    from mergify_engine import context
+
+LOG = daiquiri.getLogger(__name__)
+
+DEFAULT_VERSION = 1
+EVENTLOGS_EXPIRATION = datetime.timedelta(days=90)
+EVENTLOGS_MAX_SIZE = 100000
+
+
+class EventBase(typing.TypedDict):
+    timestamp: datetime.datetime
+
+
+class EventMetadata(typing.TypedDict):
+    pass
+
+
+class EventLabel(EventBase):
+    event: typing.Literal["action.label"]
+    metadata: signals.EventLabelMetadata
+
+
+class EventRefresh(EventBase):
+    event: typing.Literal["action.refresh"]
+    metadata: signals.EventNoMetadata
+
+
+def _get_pull_request_key(
+    owner_id: github_types.GitHubAccountIdType,
+    repo_id: github_types.GitHubRepositoryIdType,
+    pull_request: github_types.GitHubPullRequestNumber,
+) -> str:
+    return f"eventlogs/{owner_id}/{repo_id}/{pull_request}"
+
+
+Event = typing.Union[EventLabel, EventRefresh]
+
+SUPPORTED_EVENT_NAMES = list(
+    itertools.chain(
+        *[
+            evt.__annotations__["event"].__args__
+            for evt in Event.__args__  # type: ignore[attr-defined]
+        ]
+    )
+)
+
+
+class GenericEvent(EventBase):
+    event: signals.EventName
+    metadata: signals.EventMetadata
+
+
+class EventLogsSignal(signals.SignalBase):
+    async def __call__(
+        self,
+        repository: "context.Repository",
+        pull_request: github_types.GitHubPullRequestNumber,
+        event: signals.EventName,
+        metadata: signals.EventMetadata,
+    ) -> None:
+        if event not in SUPPORTED_EVENT_NAMES:
+            return
+
+        redis = repository.installation.redis.eventlogs
+        if redis is None:
+            return
+
+        if not repository.installation.subscription.has_feature(
+            subscription.Features.EVENTLOGS
+        ):
+            return
+
+        key = _get_pull_request_key(
+            repository.installation.owner_id, repository.repo["id"], pull_request
+        )
+        pipe = await redis.pipeline()
+        now = date.utcnow()
+
+        # FIXME(sileht): we need redis 6.2 and aioredis to be able set MINID option instead of MAXLEN to trim the eventlogs
+        # automatically at EVENTLOGS_EXPIRATION
+        # This will help to have a max size that depends on time and not on the number of elements for the eventlogs
+        await pipe.xadd(
```

Comment:
I don't think we want to store this forever and streams allow us to cap the size of the list with dates instead of the number of elements without the need to scan the redis database to find data to delete.

Reply:
I'm not sure using time rather than the number ofelements is a good design choice.

If your typical retention is 7 days, but your PR has no activity for 10 days, you'd be still happy to see something about what happened 10 days ago — and see that no activity happened since then (rather than seeing "nothing" which can make you doubt about things even working).

Reply:
The reverse is also true, I think the best is maybe to have something like:
* you keep at least 7 days
* you keep at least X items

Handling both criteria means we may need manual cleanup of the structure whatever we choose LIST to STREAM.

Reply:
Agreed, though it seems way easier to manipulate and trim lists in that regard than streams, which are built for an entirely different purpose.

Also I don't think this is going to be "keep at least XXX" (especially for time-based criteria) as that does not cap the storage and that's not what we want to provide — infinite storage events.